server/controller/src/main/java/ai/starwhale/mlops/common/KeyLock.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright 2022 Starwhale, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ai.starwhale.mlops.common;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+public class KeyLock<T> implements Lock {
+    private static final ConcurrentHashMap<Object, LockAndCounter> locksMap = new ConcurrentHashMap<>();
+
+    private final T key;
+
+    public KeyLock(T lockKey) {
+        this.key = lockKey;
+    }
+
+    private static class LockAndCounter {
+        private final Lock lock = new ReentrantLock();
+        private final AtomicInteger counter = new AtomicInteger(0);
```

Comment:
if no read write use cases, what is counter used for

Reply:
used for clean the map